Enable DMA for NXP LPSPI Driver

danieldegrasse opened the pull request:
This PR enables SPI DMA on all NXP RT10xx platforms, which utilize the LPSPI driver. DMA support should work on other SOCs that have the LPSPI peripheral, but is not currently tested nor enabled in those devicetrees. This PR enables SPI DMA on the following evaluation boards:

- RT1064
- RT1060
- RT1050
- RT1024
- RT1020
- RT1015
- RT1010

mmahadevan108 commented on tests/drivers/spi/spi_loopback/src/spi.c:
```diff
@@ -35,15 +35,27 @@ struct spi_cs_control spi_cs = {
 uint8_t buffer_tx[] = "0123456789abcdef\0";
-uint8_t buffer_rx[BUF_SIZE] = {};
+uint8_t __aligned(16) buffer_rx[BUF_SIZE] = {};
 
```
Why are these changes made to the `#else` case?

danieldegrasse replied:
Those changes were made to ensure that the rx buffer would be aligned for DMA controllers that may require it, as DMA buffers are aligned in the DMA tests in Zephyr. After testing, the EDMA controller in the RT10xx series does not require aligned buffers, so this change has been removed

mmahadevan108 commented on dts/arm/nxp/nxp_rt1010.dtsi:
```diff
@@ -68,6 +68,10 @@
+	dma-channels = <16>;
+};
+
 / {
```
Can this move to a separate commit with details on why this change is needed.

danieldegrasse replied:
Done, the RT1010 has only 16 channels on its EDMA while other RT10xx platforms have 32.

mmahadevan108 commented on boards/arm/mimxrt1064_evk/mimxrt1064_evk.dts:
```diff
@@ -250,8 +250,14 @@ zephyr_udc0: &usb1 {
 	status = "okay";
+	/* DMA channels 2 and 3, muxed to LPSPI3 RX and TX */
+	dmas = <&edma0 2 15>, <&edma0 3 16>;
+	dma-names = "rx", "tx";
```
Should changes to this file move the commit related to changes to the `boards` folder

danieldegrasse replied:
Fixed, thank you

danieldegrasse committed:
drivers: spi_mcux_lpspi: Added DMA support to NXP LPSPI driver

Adds DMA support to NXP's LPSPI driver. This can be enabled by selecting
the KConfig symbol CONFIG_SPI_MCUX_LPSPI_DMA, and requires the LPSPI
instances enabled in the devicetree to have valid DMA instances
assigned.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
dts: mimxrt1010: Correct DMA channel count for RT1010 SOC

RT1010 SOC has 16 DMA channels on its EDMA controller. Update
devicetree to reflect this.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt10xx: Added support for LPSPI DMA on RT10xx series

Add support for LPSPI DMA mode on RT1064, RT1060, RT1050, RT1024,
RT1020, RT1015, and RT1010 evaluation boards. Update tests to match.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

mmahadevan108 commented on drivers/spi/spi_mcux_lpspi.c:
```diff
@@ -30,13 +32,38 @@ struct spi_mcux_config {
+#endif
 };
 
+
```
Minor point, this extra newline can be deleted

mmahadevan108 commented on drivers/spi/spi_mcux_lpspi.c:
```diff
@@ -197,6 +224,232 @@ static int spi_mcux_configure(const struct device *dev,
+	int ret;
+	size_t dma_size;
+
+
```
Minor point, one of the new lines can be deleted

mmahadevan108 commented:
@danieldegrasse, could you help resolve the CI build errors.

cfriedt commented:
@danieldegrasse - can you rebase / resolve merge conflicts?

danieldegrasse commented:
@cfriedt Done, thanks for the heads up

danieldegrasse commented:
@bbilas Could you take a look at the CI failure? When I bisect it the commit that introduced the failure is 97d3de23. The build failure results from building `tests/drivers/spi/spi_loopback` targeting the `cy8ckit_062_ble_m0` board.

bbilas commented:
> @bbilas Could you take a look at the CI failure? When I bisect it the commit that introduced the failure is [97d3de2](https://github.com/zephyrproject-rtos/zephyr/commit/97d3de2349500935abbc5e69d140a403cf734208). The build failure results from building `tests/drivers/spi/spi_loopback` targeting the `cy8ckit_062_ble_m0` board.

These 2 MRs should fix those failures:

https://github.com/zephyrproject-rtos/zephyr/pull/40225
https://github.com/zephyrproject-rtos/zephyr/pull/40217

cfriedt commented:
> > @bbilas Could you take a look at the CI failure? When I bisect it the commit that introduced the failure is [97d3de2](https://github.com/zephyrproject-rtos/zephyr/commit/97d3de2349500935abbc5e69d140a403cf734208). The build failure results from building `tests/drivers/spi/spi_loopback` targeting the `cy8ckit_062_ble_m0` board.
> 
> These 2 MRs should fix those failures:
> 
> #40225 #40226

@bbilas - you could combine them into a single PR. Might be preferable.

bbilas commented:
> > > @bbilas Could you take a look at the CI failure? When I bisect it the commit that introduced the failure is [97d3de2](https://github.com/zephyrproject-rtos/zephyr/commit/97d3de2349500935abbc5e69d140a403cf734208). The build failure results from building `tests/drivers/spi/spi_loopback` targeting the `cy8ckit_062_ble_m0` board.
> > 
> > 
> > These 2 MRs should fix those failures:
> > #40225 #40226
> 
> @bbilas - you could combine them as separate commits in a single PR. Might be preferable.

I will do that.

bbilas commented:
@cfriedt @danieldegrasse https://github.com/zephyrproject-rtos/zephyr/pull/40241

cfriedt commented:
@danieldegrasse - please rebase on a recent `main`. There were some CI changes that went in recently.
